src/clients/Client.cc

```diff
@@ -820,24 +817,29 @@ Client::handleAdaptedBodyProductionEnded()
     // distinguish this code path from handleAdaptedBodyProducerAborted()
     receivedWholeAdaptedReply = true;
 
-    // end consumption if we consumed everything
-    if (adaptedBodySource != nullptr && adaptedBodySource->exhausted())
-        endAdaptedBodyConsumption();
-    // else resumeBodyStorage() will eventually consume the rest
+    checkAdaptationCompletion();
 }
 
 void
-Client::endAdaptedBodyConsumption()
+Client::checkAdaptationCompletion()
 {
-    stopConsumingFrom(adaptedBodySource);
+    if (adaptedBodySource == nullptr)
+        return;
 
-    if (receivedWholeAdaptedReply) {
-        // We received the entire adapted reply per receivedWholeAdaptedReply.
-        // We are called when we consumed everything received (per our callers).
-        // We consume only what we store per handleMoreAdaptedBodyAvailable().
-        fwd->markStoredReplyAsWhole("received,consumed=>stored the entire RESPMOD reply");
+    if (!adaptedBodySource->exhausted()) {
+        debugs(11,5, "waiting to consume the remainder of the adapted body");
+        return; // resumeBodyStorage() should eventually consume the rest
     }
 
+    if (!receivedWholeAdaptedReply && !adaptedReplyAborted) {
+        // wait for noteBodyProductionEnded() or noteBodyProducerAborted()
+        // because we need to know whether we receivedWholeAdaptedReply
+        return;
+    }
+
+    stopConsumingFrom(adaptedBodySource);
+
+    Assure(doneWithAdaptation());
     handleAdaptationCompleted();
```

Comment:
I think you are right, but there is some clean up code in Client::cleanAdaptation(), called from handleAdaptationCompleted(), assuming that it can be !doneWithAdaptation(). Probably this general cleanup should work from swanSong().